Sources/NIOHTTP1/HTTPObjectAggregator.swift

```diff
@@ -0,0 +1,422 @@
+//===----------------------------------------------------------------------===//
+//
+// This source file is part of the SwiftNIO open source project
+//
+// Copyright (c) 2020 Apple Inc. and the SwiftNIO project authors
+// Licensed under Apache License v2.0
+//
+// See LICENSE.txt for license information
+// See CONTRIBUTORS.txt for the list of SwiftNIO project authors
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+//===----------------------------------------------------------------------===//
+
+import NIO
+
+/// The parts of a complete HTTP response from the view of the client.
+///
+/// Afull HTTP request is made up of a response header encoded by `.head`
+/// and an optional `.body`.
+public struct HTTPServerRequestFull {
+    public var head: HTTPRequestHead
+    public var body: ByteBuffer?
+
+    public init(head: HTTPRequestHead, body: ByteBuffer?) {
+        self.head = head
+        self.body = body
+    }
+}
+
+extension HTTPServerRequestFull: Equatable {}
+
+/// The parts of a complete HTTP response from the view of the client.
+///
+/// Afull HTTP response is made up of a response header encoded by `.head`
+/// and an optional `.body`.
+public struct HTTPClientResponseFull {
+    public var head: HTTPResponseHead
+    public var body: ByteBuffer?
+
+    public init(head: HTTPResponseHead, body: ByteBuffer?) {
+        self.head = head
+        self.body = body
+    }
+}
+
+extension HTTPClientResponseFull: Equatable {}
+
+public enum HTTPObjectAggregatorError: Error {
+    case frameTooLong
+    case connectionClosed
+    case messageDropped
+    case channelStateError(event: String, state: String)
+    case finishedIgnoringContent
+}
+
+extension HTTPObjectAggregatorError: Equatable {}
+
+public enum HTTPObjectAggregatorEvents {
+    case httpExpectationFailedEvent
+    case httpContinueEvent
+    case httpFrameTooLongEvent
+}
```

Comment:
The same is true here: we probably want to make this a struct that hides an `enum`. The pattern here would be something like this:

```swift
public struct HTTPObjectAggregatorEvent: Hashable {
    private enum Base {
        case httpExpectationFailedEvent
        case httpContinueEvent
        case httpFrameTooLongEvent
    }

    private var base: Base

    private init(base: Base) {
        self.base = base
    }

    public static let httpExpectationFailedEvent = HTTPObjectAggregatorEvent(base: . httpExpectationFailedEvent)
    public static let httpContinueEvent = HTTPObjectAggregatorEvent(base: . httpContinueEvent)
    public static let httpFrameTooLongEvent = HTTPObjectAggregatorEvent(base: . httpFrameTooLongEvent)
}
```

This somewhat awkward pattern is necessary because you cannot add new cases to Swift `enum`s without breaking existing source. The same is not true for the above, which is essentially an "open" enumeration.

While we're here, the word `event` is redundant and so we can probably remove it from the enum cases and `static let`s. It's in the type, so that's sufficient.

Reply:
used the pattern for both and made the error throwing much more verbose (including bubbling everything up via `fireErrorCaught`) - not sure if not too verbose though?

Reply:
I think it's fine: it's the price we have to pay.